src/support/rand.c

```diff
@@ -28,43 +28,68 @@
 
 #include "wt_internal.h"
 
+/*
+ * This is an implementation of George Marsaglia's multiply-with-carry pseudo-
+ * random number generator.  Computationally fast, with reasonable randomness
+ * properties.
+ *
+ * Be very careful about races here. Multiple threads can call __wt_random
+ * concurrently, and it is okay if those concurrent calls get the same return
+ * value. What is *not* okay is if reading/writing the shared state races and
+ * uses two different values for m_w or m_z. That can result in a stored value
+ * of zero, in which case they will be stuck on zero forever. Take a local copy
+ * of the values to avoid that, and read/write in atomic, 8B chunks.
+ */
 #undef	M_W
-#define	M_W	(rnd)[0]
+#define	M_W(p)	((uint32_t *)&(p))[0]
 #undef	M_Z
-#define	M_Z	(rnd)[1]
+#define	M_Z(p)	((uint32_t *)&(p))[1]
 
 /*
  * __wt_random_init --
  *	Initialize return of a 32-bit pseudo-random number.
  */
 void
-__wt_random_init(uint32_t *rnd)
+__wt_random_init(uint64_t * volatile rnd_state)
 {
-	M_W = 521288629;
-	M_Z = 362436069;
+	uint64_t rnd;
+
+	M_W(rnd) = 521288629;
+	M_Z(rnd) = 362436069;
+	*rnd_state = rnd;
 }
 
 /*
  * __wt_random --
  *	Return a 32-bit pseudo-random number.
- *
- * This is an implementation of George Marsaglia's multiply-with-carry pseudo-
- * random number generator.  Computationally fast, with reasonable randomness
- * properties.
- *
- * We have to be very careful about races here.  Multiple threads can call
- * __wt_random concurrently, and it is okay if those concurrent calls get the
- * same return value.  What is *not* okay is if reading the shared state races
- * with an update and uses two different values for m_w or m_z.  That could
- * result in a value of zero, in which case they would be stuck on zero
- * forever.  Take local copies of the shared values to avoid this.
  */
 uint32_t
-__wt_random(uint32_t *rnd)
+__wt_random(uint64_t * volatile rnd_state)
```

Comment:
The type annotation is wrong. This means that that you have a volatile pointer to non-volatile memory. Since it is the memory pointed to by `rnd_state` that is volatile, you need to write as `volatile uint64_t* rnd_state` or `uint64_t volatile * rnd_state`.